DakshChan/IBS
# [IBS-44] Add task weight column + accumulated weight check

## Overview

[Jira Ticket Link](https://ibs-utm.atlassian.net/jira/software/projects/IBS/boards/1?selectedIssue=IBS-44)

- Add `weight` column to `task` table.
- Ensure all endpoints using `task` table is changed appropriately. i.e. return `weight` property if using `GET` request, validate whether `weight` property exists on `POST, PUT` requests.
- When adding a new task, validate whether `sum(all previous tasks' weights) + new task weight <= 100`. Otherwise, if the accumulated weights exceed 100, return `400` with appropriate message.
- In addition, because `AggregatedGrades` page on frontend needs the task weight property from calling `/api/instructor/course/{{course_id}}/mark/all`, the said endpoint should also return the task weight property.
  e.g. a response from the above endpoint should look like the following:
  ```json
  {
      "marks": {
          "student1": {
              "task7": {
                  "mark": 4,
                  "out_of": 42,
                  "weight": 81
              }
          }
      }
  }
  ```

## Testing
- Make sure database is reset. To do so, you can just purge all docker containers by using the command `docker system prune --all --force --volumes`.

### Testing `task` endpoints
1. Add `admin` credentials.
2. As admin, create a new course.
3. As admin, add new role for `instructor` and add the instructor to the new course.
4. Make sure that in the DB, the `task` table for the created course contains `weight` column.
5. As instructor, test the following endpoints (ensuring `weight` key is set in the request body):
    - `GET All tasks`
    - `GET Get task`
    - `POST Add Task`
    - `PUT Change task`
6. Verify that the accumulated weight check works as intended.

### Testing `GET Get all marks` endpoint
1. Follow the above steps 1-4 if not already done.
2. As admin, add a new role for `student` and add the student to the course.
3. As instructor, add a new task by hitting the endpoint `POST Add Task`.
4. As instructor, add new criteria by hitting the endpoint `POST Add criteria`.
5. As instructor, submit a mark for a student in the course by hitting the endpoint `POST Submit mark`.
6. As instructor, test the endpoint `GET Get all marks` to verify that `weight` property is added to each task in the response.
